### PR TITLE
actionpack: signature updates

### PR DESCRIPTION
**Pull request by hjwylde**
Split from #735 to be smaller. Type signature updates for invalid signatures in actionpack, the majority are due to blocks being marked as required, but they should be optional.

### CODE REVIEW

File: gems/actionpack/6.0/actiondispatch.rbs [hunk 1]
```diff
@@ -228,6 +275,61 @@ module ActionDispatch
...
+      # for example the +path+ option in the last example.
+      # <tt>/stories</tt>, <tt>/stories?foo=bar</tt>, redirect to <tt>/posts</tt> and <tt>/posts?foo=bar</tt> respectively.
+      #
+      # Finally, an object which responds to call can be supplied to redirect, allowing you to reuse
+      # common redirect routes. The call method must accept two arguments, params and request, and return
+      # a string.
+      #
+      #   get 'accounts/:name' => redirect(SubdomainRedirector.new('api'))
+      #
+      def redirect: (*untyped args) ?{ () -> untyped } -> (OptionRedirect | PathRedirect | Redirect)
```
Comment (tk0miya) [inline gems/actionpack/6.0/actiondispatch.rbs:328]: As a quick look, the option is only needed when no arguments are given except the status option.

I'm not an expert on this API, but is it better to replace this with this?

```suggestion
      def redirect: (*untyped args) -> (OptionRedirect | PathRedirect | Redirect)
                  | (?status: untyped) { () -> untyped } -> Redirect
```

@ksss Do you have knowledge for this?

refs: https://github.com/rails/rails/blob/v6.0.6.1/actionpack/lib/action_dispatch/routing/redirection.rb#L187-L198

File: gems/actionpack/6.0/actioncontroller.rbs [hunk 2]
```diff
@@ -161,6 +161,12 @@ end
 
 module ActionController
   class Parameters
+    # Deletes a key-value pair from +Parameters+ and returns the value. If
+    # +key+ is not found, returns +nil+ (or, with optional code block, yields
+    # +key+ and returns the result). Cf. +#extract!+, which returns the
+    # corresponding +ActionController::Parameters+ object.
+    def delete: (untyped key) ?{ () -> untyped } -> untyped
```
[the remark this answers is not in the document]
Comment (hjwylde) [reply gems/actionpack/6.0/actioncontroller.rbs]: Thanks, looks like you are right, I'll add in the key parameter.

File: gems/actionpack/6.0/actioncontroller.rbs [hunk 3]
```diff
@@ -255,3 +261,17 @@ module ActionController
...
+  module ContentSecurityPolicy
+    module ClassMethods
+      def content_security_policy: (?bool enabled, **untyped options) ?{ (untyped) -> untyped } -> untyped
+    end
+  end
+end
+
+module ActionController
+  class TestSession < Rack::Session::Abstract::PersistedSecure::SecureSessionHash
+    def fetch: (untyped key, *untyped args) ?{ () -> untyped } -> untyped
```
[the remark this answers is not in the document]
Comment (hjwylde) [reply gems/actionpack/6.0/actioncontroller.rbs:275]: That looks reasonable to me, thank you. I'll make the update.

File: gems/actionpack/6.0/actiondispatch.rbs [hunk 4]
```diff
@@ -1,3 +1,50 @@
+module ActionDispatch
+  class Cookies
+    class CookieJar
+      def fetch: (untyped name, *untyped args) ?{ () -> untyped } -> untyped
```
[the remark this answers is not in the document]
Comment (hjwylde) [reply gems/actionpack/6.0/actiondispatch.rbs]: Thanks! Accepted the suggestion.

File: gems/actionpack/6.0/actiondispatch.rbs [hunk 5]
```diff
@@ -1,3 +1,50 @@
...
+    class Headers
+      # Returns the value for the given key mapped to @env.
+      #
+      # If the key is not found and an optional code block is not provided,
+      # raises a <tt>KeyError</tt> exception.
+      #
+      # If the code block is provided, then it will be run and
+      # its result returned.
+      def fetch: (untyped key) { () -> untyped } -> untyped
+               | (untyped key, untyped default) -> untyped
```
[the remark this answers is not in the document]
Comment (hjwylde) [reply gems/actionpack/6.0/actiondispatch.rbs:21]: I think you are right, but it does short circuit and never hits the block if a default is given. In terms of a signature that represents the type behaviour of the method, I think omitting the block when a default is present is more accurate; but I don't have a strong opinion on that, so let me know your preference, and I can add the optional block in for the second overload if you prefer.

File: gems/actionpack/6.0/actiondispatch.rbs [hunk 6]
```diff
@@ -1,3 +1,50 @@
...
+      #
+      #   session.fetch(:foo, :bar)
+      #   # => :bar
+      #
+      #   session.fetch(:foo) do
+      #     :bar
+      #   end
+      #   # => :bar
+      def fetch: (untyped key) { () -> untyped } -> untyped
+               | (untyped key, untyped default) -> untyped
```
[the remark this answers is not in the document]
Comment (hjwylde) [reply gems/actionpack/6.0/actiondispatch.rbs:44]: For this one I was going off the documentation:

> or raises +KeyError+ if can't find the given key and no default value is set.

There's currently no ways to mark error paths in type signatures, so this is how its written in `core/hash.rbs` as well (https://github.com/ruby/rbs/blob/1a3ab85f568d0230dffcebb96d9d0b8e15c05d79/core/hash.rbs#L1108).

File: gems/actionpack/6.0/actiondispatch.rbs (hunk 1 above)
Comment (hjwylde) [reply gems/actionpack/6.0/actiondispatch.rbs:329]: I think that looks accurate, except that the overloads should be written in the opposite order, as I believe priority is given to the first overload during type checking. I'll make the change with this fix for now, but we can always revert/update based on further advice.

File: gems/actionpack/6.0/actiondispatch.rbs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tk0miya) [reply gems/actionpack/6.0/actiondispatch.rbs:21]: Ah, sorry. My last comment is wrong. I wanted to say "when the default parameter is NOT given."

If my understanding is correct, we can call `#fetch` with key parameter (without block).
Therefore it's better to define this method as the followings (either is okay to me).

```
      def fetch: (untyped key) -> untyped
               | (untyped key, untyped default) -> untyped
               | (untyped key) { () -> untyped } -> untyped
```
```
      def fetch: (untyped key) ?{ () -> untyped } -> untyped
               | (untyped key, untyped default) -> untyped
```

File: gems/actionpack/6.0/actiondispatch.rbs (hunk 6 above)
[the remark this answers is not in the document]
Comment (hjwylde) [reply gems/actionpack/6.0/actiondispatch.rbs:44]: Oh, I misunderstood the comment. Yes I guess the block is optional, and the KeyError is then thrown if no default value and no block is given. Will fix, thank you.

File: gems/actionpack/6.0/actiondispatch.rbs (hunk 5 above)
[the remark this answers is not in the document]
Comment (hjwylde) [reply gems/actionpack/6.0/actiondispatch.rbs:21]: Thanks! Yes you are right, I seem to have got my head in a muddle and thought it was mandatory for some reason, but the block is optional. Will update.

**Review by tk0miya — COMMENTED** [review]
Thank you for update. looks good.

APPROVE

**Pull request merged**